copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+    public static String TASK_KEY(ConnectorTaskId taskId) {
+        return TASK_PREFIX + taskId.connector() + "-" + taskId.task();
+    }
+
+    // Note that while using real serialization for values as we have here, but ad hoc string serialization for keys,
+    // isn't ideal, we use this approach because it avoids any potential problems with schema evolution or
+    // converter/serializer changes causing keys to change. We need to absolutely ensure that the keys remain precisely
+    // the same.
+    public static final Schema ROOT_V0 = SchemaBuilder.struct()
+            .field("connectors", SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.INT32_SCHEMA).build())
+            .build();
+    public static final Schema CONNECTOR_CONFIGURATION_V0 = SchemaBuilder.struct()
+            .field("properties", SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.OPTIONAL_STRING_SCHEMA))
+            .build();
+    public static final Schema TASK_CONFIGURATION_V0 = CONNECTOR_CONFIGURATION_V0;
+
+    private static final long READ_TO_END_TIMEOUT_MS = 30000;
+
+    private final Object lock;
+    private boolean starting;
+    private final Converter converter;
+    private final Callback<String> connectorConfigCallback;
+    private final Callback<List<ConnectorTaskId>> tasksConfigCallback;
+    private String topic;
+    // Data is passed to the log already serialized. We use a converter to handle translating to/from generic Copycat
+    // format to serialized form
+    private KafkaBasedLog<String, byte[]> configLog;
+    // Root config: connector -> # of tasks
+    private Map<String, Integer> rootConfig = new HashMap<>();
+    // Connector and task configs: name or id -> config map
+    private Map<String, Map<String, String>> connectorConfigs = new HashMap<>();
+    private Map<ConnectorTaskId, Map<String, String>> taskConfigs = new HashMap<>();
+
+    private Map<ConnectorTaskId, Map<String, String>> deferredTaskUpdates = new HashMap<>();
+
+
+    public KafkaConfigStorage(Converter converter, Callback<String> connectorConfigCallback, Callback<List<ConnectorTaskId>> tasksConfigCallback) {
+        this.lock = new Object();
+        this.starting = false;
+        this.converter = converter;
+        this.connectorConfigCallback = connectorConfigCallback;
+        this.tasksConfigCallback = tasksConfigCallback;
+    }
+
+    public void configure(Map<String, ?> configs) {
+        if (configs.get(CONFIG_TOPIC_CONFIG) == null)
+            throw new CopycatException("Must specify topic for Copycat connector configuration.");
+        topic = (String) configs.get(CONFIG_TOPIC_CONFIG);
+
+        Map<String, Object> producerProps = new HashMap<>();
+        producerProps.putAll(configs);
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer");
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
```

Comment:
Since we made `Converter` handle the complete translation from Copycat -> byte[], we don't have internal serde settings, only internal converter. In a bunch of the other code this actually got rid of all the generic types since we didn't have intermediate types that needed to be generic (i.e. we don't need to parameterize for SpecificRecord, JsonNode, etc. since Copycat never sees those intermediate types).

In this case, the class is actually instantiated like this, passing in the internal value serializer:

```
this.configStorage = new KafkaConfigStorage(worker.getInternalValueConverter(),
                new ConnectorConfigCallback(), new TaskConfigCallback());
```

The key serializer is an interesting one. We could create a key schema and use the converters for that as well. Right now it's an ad hoc string-based approach since there are a very limited number of configs and it's easy to avoid conflicts for something internal like this. One of the drawbacks of using the converter approach is that any small, even accidental, change to the serialization could break things (e.g. spacing in JSON changes, order of fields changes, etc). Of course we already have the same problem with offsets, so maybe it's not relevant. I'm open to using either approach -- the scope in this case is very limited so I'm much less concerned about the more ad hoc (and easier for humans to decode) approach.

As for acks, I've added that config.